### ISSUE TITLE
Display layout not respected during session login

### ISSUE THREAD

**Issue by XV-02** [pop-os/cosmic-greeter#21]
On a system with an internal display (eDP-1) and an external display (DP-1) such that the two displays are aligned at their bottoms with eDP-1 at the left edge of DP-1, the session starts with displays correct, and maintains display layout in the lock-screen. 

But, when first logging in, the displays are such that they are top edge aligned with eDP-1 to the right of DP-1. This was the default layout that was generated the first time that DP-1 was attached to the system. However, across many unplug-replug cycles, the displays have consistently resumed my user-set arrangement everywhere except for initial session login.

**Comment by WatchMkr** [comment]
This applies to a landscape and portrait display setup as well.

**Comment by Drakulix** [comment]
Problem is, that the configuration is user-specific, so any changes don't affect cosmic-greeter.

We could try to apply the user's config through the greeter-daemon, but that would potentially cause a modeset on user-selection, which is less than desirable.

Perhaps we can add a button to cosmic-settings to "export the display configuration to the login screen"?

The clear solution for orientation and scale would obviously be to derive sensible values properly on first startup. That doesn't help with the layout though.

(Note GNOME has a similar problem with GDM, various guides still recommend to copy the monitors.xml from a user-session to gdm.)

**Comment by Kartonrealista** [comment]
@Drakulix 
>Problem is, that the configuration is user-specific, so any changes don't affect cosmic-greeter.

They should!

> We could try to apply the user's config through the greeter-daemon, but that would potentially cause a modeset on user-selection, which is less than desirable.
> Perhaps we can add a button to cosmic-settings to "export the display configuration to the login screen"?

From a user's perspective it seems preferable to use the last user's last display config by default than do anything else, since most people use one user to begin with. Things should work by default instead of being broken by default for most people, and then you get an option to maybe fix it in the settings. Maybe this would be better as a checkbox instead of a button, and it's ticked by default. As long as it would be ticked, the settings would be exported when the display is configured. 

The multiple users edge case seems almost irrelevant. Even if multiple people use the same computer, they have monitors arranged in a particular way on their desk and that won't magically change upon login. So configuration being user specific seems pointless to begin with.

>The clear solution for orientation and scale would obviously be to derive sensible values properly on first startup.

This is hardly clear at all, especially if you have to fix the resolution on one of your monitors because it's old or something and it doesn't get detected properly. Or change the orientation. First startup permaborking cosmic-greeter unless you dig into config files is a less than welcome proposition. 

Especially if you change your layout all the time. I use one of my monitors in a vertical orientation when viewing vertical images (like when reading comics) and horizontal otherwise. It would be so much less hassle if the last config used was preserved and if logging in next time remembered what you set. Clicking a button every time so it doesn't break seems supremely annoying.

**Comment by Drakulix** [comment]
> From a user's perspective it seems preferable to use the last user's last display config by default than do anything else, since most people use one user to begin with.

Yes, though I disagree that this should sync back into a user's configuration, which your further statements seem to suggest.

> The multiple users edge case seems almost irrelevant. Even if multiple people use the same computer, they have monitors arranged in a particular way on their desk and that won't magically change upon login.

This doesn't only include orientation, layout and resolution, but also settings like mirroring or scale. Here it is pretty obvious, that different users *might* prefer different settings, even *if* admittedly unlikely.

> This is hardly clear at all, especially if you have to fix the resolution on one of your monitors because it's old or something and it doesn't get detected properly.

Which is a reason why I think a configuration sync'd to cosmic-greeter, should be sync to users on *first* login. But not afterwards. If they update their config, it should stay user specific. (And only sync back to the greeter.)

> I use one of my monitors in a vertical orientation when viewing vertical images (like when reading comics) and horizontal otherwise.

Note that many monitors report their current orientation to the system. So in an ideal scenario, it would update the settings automatically upon rotation. But again, if we sync back to cosmic-greeter, the last configuration would stick, which does make sense.

**Comment by Kartonrealista** [comment]
>Which is a reason why I think a configuration sync'd to cosmic-greeter, should be sync to users on first login. But not afterwards. If they update their config, it should stay user specific. (And only sync back to the greeter.)

I think I misunderstood what you said in the previous comment. I thought you said the greeter would inherit first login config from the desktop. If it's the opposite that's reasonable. I just wish changing your desktop settings synced back to the greeter by default, and this could be turned off if it bothered someone. I don't think the greeter settings should sync into all users' configuration.

>This doesn't only include orientation, layout and resolution, but also settings like mirroring or scale. Here it is pretty obvious, that different users might prefer different settings, even if admittedly unlikely.

Point taken.

**Comment by aerialviews007** [comment]
Where this issue comes into play is where the laptop hardware's default display orientation does not match the intended usage.  This is my case with the GPD Pocket 3.  Normally, a change in the kernel parameter (fbcon=rotate:1 video=DSI-1:panel_orientation=right_side_up) fixes the issue but cosmic-greeter does not respect this setting.

**Comment by universal** [comment]
another issue regarding this: I have a laptop + two external displays. I have the laptop display configured to be off. The greeter leaves it on (which is okayish), but: Once I login the display gets only partially deactivated, and still shows some log messages after the desktop has fully loaded. The laptop display is shown as off in the settings dialog, and toggling it on/off there fixes the problem. 

Might be a separate bug, but it feels connected to this.
